File: LeetCode/Python/0017. Letter Combinations of a Phone Number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []
        
        all_combinations = []
        self._letterCombinations(digits, [], all_combinations)
        return all_combinations
    
    def getMapping(self, n):
        return ("", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz")[ord(n) - ord('0')]
    
    def _letterCombinations(self, digits, current, all_combinations):
        if not digits:
            all_combinations.append("".join(current))
            return
        
        mapping = self.getMapping(digits[0])
        for char in mapping:
            self._letterCombinations(digits[1:], current + [char], all_combinations)

File: LeetCode/Java/0017. Letter Combinations of a Phone Number.java
class Solution {
    public List<String> letterCombinations(String digits) {
        var allCombinations = new ArrayList<String>();
        
        if (digits.length() == 0)
            return allCombinations;
        
        letterCombinations(digits, "", allCombinations);
        return allCombinations;
    }
    
    private void letterCombinations(String digits, String current, ArrayList<String> allCombinations) {
        if (digits.length() == 0) {
            allCombinations.add(current);
            return;
        }
        
        String mapping = getMapping(digits.charAt(0));
        for (int i = 0; i < mapping.length(); i++)
            letterCombinations(digits.substring(1), current + mapping.charAt(i), allCombinations);
    }
    
    private String getMapping(char n) {
        return new String[] {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"} [n - '0'];
    }
}
